qt-cpp: add build test

lugerard opened the pull request:
- Add integration test that configures and builds a minimal Qt app using CMake Tools with CMAKE_PREFIX_PATH taken from qt-core.qtInstallationRoot
- ProjectFolder is copied to a temp dir before each run to isolate state
- Temp dir and build artifacts are cleaned up automatically afterwards
- Run works both in VSCode (via tasks and script) and terminal (via runTest.build.mts)

<!---
# Qt contribution guidelines

We welcome contributions to Qt!

Read the
[Qt Contribution Guidelines](https://wiki.qt.io/Qt_Contribution_Guidelines) to learn more.
<!---

lugerard committed:
qt-cpp: add build test

- Add integration test that configures and builds a minimal Qt app
  using CMake Tools with CMAKE_PREFIX_PATH taken from qt-core.qtInstallationRoot
- ProjectFolder is copied to a temp dir before each run to isolate state
- Temp dir and build artifacts are cleaned up automatically afterwards
- Run works both in VSCode (via tasks and script) and terminal (via runTest.build.mts)

lugerard committed:
package: license

lugerard committed:
github: removing trailing spaces

lugerard committed:
qt-cpp: Correct test script launch

The qt-root arg is now propagated to both runTest and runTest.build launches.

OrkunTokdemir commented on package.json:
```diff
@@ -129,6 +129,7 @@
   },
   "devDependencies": {
     "@eslint/js": "^9.21.0",
+    "@types/commander": "^2.12.0",
```
Where do you use `commander`?

lugerard replied:
removed

OrkunTokdemir commented on .vscode/launch.json:
```diff
@@ -25,6 +25,22 @@
+        "--disable-restore-windows"
+      ],
+      "outFiles": ["${workspaceFolder}/qt-cpp/out/test/**/*.js"],
+      "preLaunchTask": "qt-cpp: build test prelaunch chain",
```
I think we have separated our test code into `test` and `scripts` currently. What we have done in `scripts/clean_test_build.ts` and `scripts/prepare_test_build.ts` can be implemented with a default environment class in our test folder. For example, https://github.com/microsoft/vscode-cmake-tools/blob/main/test/helpers/test/default-environment.ts#L12. The main idea of the `scripts` folder is to manage the project easily with fewer CLI commands, like generating licenses, starting tests, publishing, etc. Having test-specific code is beyond the scope of that folder.

lugerard replied:
The idea was to have the test run on clean temp workspace. In order to do that, the workspace needs to be created before the test is launched, before VSCode host is launched, so this must happen in runTest. runTest is not used when launching the test from VSCode as opposed to launching it from the terminal. Those scripts are here only so that running the test in VSCode behaves the same as running the test from the terminal. That being said, I can create a qt-cpp/test/script and move those scripts there.

lugerard committed:
qt-cpp: Build test related script moved to qt-cpp/test/script folder

lugerard committed:
flow: Introduce QT_VERSION_FOR_TEST

lugerard committed:
qt-cpp:build-test: Set QT_TEST_QT_ROOT from qt-core.qtInstallationRoot

lugerard committed:
qt-cpp:build-test: Correct path in test scripts

lugerard committed:
qt-cpp:build-test: Take explicit control of Qt6_DIR

Qt6_DIR is used by cmake.
Qt6_DIR is set up automatically by install-qt-action,
we make sure it is set to something within the QT_TEST_QT_ROOT.
If not set, we set it up to the correct leaf within QT_TEST_QT_ROOT.
Qt version can be provided through QT_VERSION_FOR_TEST, otherwise
it defaults to 6.9.0 or lower versions.

lugerard committed:
flow: Change qt version for test

back to 6.6.0

lugerard committed:
qt-cpp:build-test: Refactor build related functionalities

lugerard committed:
flow: Fix Qt installation after qt-ui testing
